vm/devices/net/net_mana/src/lib.rs

```diff
@@ -1118,18 +1119,94 @@ impl<T: DeviceBacking> ManaQueue<T> {
                 1
             };
 
+            let mut bounced_len_with_padding_total = bounced_len_with_padding;
             let segment_count = tail_sgl_offset + meta.segment_count - header_segment_count;
-            let sgl = &mut sgl[..segment_count];
-            for (tail, sge) in segments[header_segment_count..]
-                .iter()
-                .zip(&mut sgl[tail_sgl_offset..])
-            {
-                *sge = Sge {
-                    address: self.guest_memory.iova(tail.gpa).unwrap(),
-                    mem_key: self.mem_key,
-                    size: tail.len,
-                };
-            }
+            let sgl = if segment_count < 31 {
+                let sgl = &mut sgl[..segment_count];
+                for (tail, sge) in segments[header_segment_count..]
+                    .iter()
+                    .zip(&mut sgl[tail_sgl_offset..])
+                {
+                    *sge = Sge {
+                        address: self.guest_memory.iova(tail.gpa).unwrap(),
+                        mem_key: self.mem_key,
+                        size: tail.len,
+                    };
+                }
+                sgl
+            } else {
+                // Tx segment count exceeds the limit. Coalescing using a bounce buffer.
+                let mut segment_index = header_segment_count;
+                let mut segments_remaining = segment_count;
+                let mut segments_peekable = segments[header_segment_count..].iter().peekable();
+                let mut sgl_index = tail_sgl_offset;
+                while segments_remaining >= 31 && segments_peekable.peek().is_some() {
+                    let seg = segments_peekable.next().unwrap();
```

Comment:
Don't use peek.is_some() + next().unwrap() like this. Just do `let Some(seg) = segments.next() else { break }`